_posts/2019-05-27-Spring-Data-Jdbc.md

```diff
@@ -0,0 +1,676 @@
+
+{% highlight java %}
+
+    public class CarRowMapper implements RowMapper<Car> {
+    
+        @Override
+        public Car mapRow(ResultSet resultSet, int rowNumber) throws SQLException {
+            Car car = new Car();
+     
+            car.setId(resultSet.getInt("ID"));
+            car.setColor(resultSet.getString("COLOR"));
+            car.setBrand(resultSet.getString("BRAND"));
+            car.setModel(resultSet.getString("MODEL"));
+     
+            return car;
+        }
+    }
+    
+{% endhighlight %}
+
+This mapper can be passed to the jdbc template so the conversion can be used to get populated java objects.
+
+{% highlight java %}
+
+    List<Car> cars = jdbcTemplate.queryForObject(
+        "SELECT * FROM CAR WHERE ID = ?", new Object[] {id}, new CarRowMapper());
+    
+{% endhighlight %}
+
+#### Spring data jpa
+
+When you use the spring data framework, it will help you with building your queries and getting the right information.
+The spring data jpa framework uses hibernate. They make it possible to query the database using user friendly interfaces.
+When you want to query the database, instead of writing the entire query yourself, hibernate will help you. 
+There are multiple ways to query the database using Spring data jpa, but they all need you to extend the repository of the entity you want to query.
+
+Some basic queries can be written using derived queries. An example of this is findById. 
+For these methods spring data will generate the sql entirely on its own.
+
+If you need to write more advanced queries that can't easily be defined as a derived query, you can define the query yourself using the @Query annotation.
+Inside the @Query annotation you write jpql or sql statements. Jpql is an SQL like syntax which provides a layer on top regular sql. 
+When jpql is used, it is possible for spring data to help you with handling the data. For example paging and sorting can be done by simply adding a parameter.
+
+If you want to be a bit more in control, you can use sql by setting the the property native = true. 
+Then you don't use this extra layer, but it then also can't help you anymore.
+
+Apart from helping you with more easily defining which query you want to execute, it also helps you with fine tuning performance.
+You are using entities when you query using spring data jpa. These entities have connections to other entities.
+Spring data jpa will help you with defining if you want to return these connected entities directly or if you don't want to do that. 
+It can help you with searching for these entities when you do need them.
+This is called eager and lazy loading and this can all be managed by spring data jpa.
+
+It will also try to improve performance by giving you the option to cache the results of these queries.
+If you would have the same query, this cache can return the result instead of letting the database calculate it again.
```

Comment:
I have rewritten the sentence. Hopefully this makes it more clear that it will not return always the same result. It will try to use it when the data hasn't changed. Can you see if it is ok now?